### DISCUSSION TITLE
Different values ​​in date fields

### DISCUSSION THREAD

**Discussion by hugopvl** [directus/directus#7350] [Archive]
good morning guys,

I'm confused by this situation, I was waiting for the rc.89 version because it fixed the tz problem with datetime, but in the meantime I did a test here and created a new rc89 installation with sql database. so I created a different type:

Server tz= GMT+1 Because im in portugal

In derectus stay:
Date: minus one day;
Timestamp: minus one hour;
![image](https://user-images.githubusercontent.com/81326296/129006548-81079608-8063-4186-91e1-39f5620c2bc6.png)

Then i change the server to TZ= Europe/Lisbon = WEST

its the same:

![image](https://user-images.githubusercontent.com/81326296/129007272-c06318f8-c843-435d-bdc2-f2d13dfd7267.png)

how can i resolve the date issue!

I had to put the server in UTC, the day is correct but it's less than 1h than in Portugal, I have to choose, have the correct day or time :) any solution ?

![image](https://user-images.githubusercontent.com/81326296/129009639-3c4bf1c2-64b4-46c7-a3e5-fe6a219cde30.png)

**Comment by hugopvl** [comment]
type DateTime: SetNow works nice, but when its automatic (Save Currente Date/Time)
its diferente:


![image](https://user-images.githubusercontent.com/81326296/129348335-a291d95f-d28c-4726-85f2-4bbaf89adb1c.png)
Teste3- Type DateTime: SetNow
Update2 - Type DateTime: (Save Currente Date/Time)
![image](https://user-images.githubusercontent.com/81326296/129348456-dc7d4697-a2bc-4308-acbc-66d79e45a323.png)

is it supposed to be this way?
and sorry for all my questions i'm still new to this stuff :D

Thanks Guys.

**Reply by rijkvanzanten** [reply]
Nope!

Just to confirm, is this a `datetime` field or a `timestamp` field? What database are you using?

Some of the DB vendors are doing some timezone magic under the hood to be "helpful", which can cause some issues when the server/database time is custom (or different from one another)

**Reply by rijkvanzanten** [reply]
See https://github.com/directus/directus/issues/7383 as well

**Comment by hugopvl** [comment]
The tow fields is datetime - SQLSERVER 2014 sp3

![image](https://user-images.githubusercontent.com/81326296/129366569-f83f7d92-9078-4dd0-8be7-167a04fad627.png)

![image](https://user-images.githubusercontent.com/81326296/129366606-d0517d6e-5ea9-44bb-bbe6-f95a8bd05d84.png)

![image](https://user-images.githubusercontent.com/81326296/129366686-847c6b01-bee4-4bfa-a16a-e2edb9db85cb.png)

![image](https://user-images.githubusercontent.com/81326296/129366760-721040ae-befd-498b-a49e-20d220b61abc.png)

**Reply by rijkvanzanten** [reply]
You mentioned your server timezone is +1, what is the timezone of your MS SQL server time?

**Reply by hugopvl** [reply]
It's not on my server, but in portugal right now it's UTC+1, daylight saving time

Querying directly in sql is: GMT standard Time

I also had this problem, #7383  as well, and my solution was to put the server in "set-timezone UTC", because previously it was like europe/lisbon - WEST, and that way it gave the problem... what happens is that working in UTC the day is correct but time is not.

But what it seems to me is that a correction was made to the datetime because the setnow was different from the time of the DB correct?

then datetime (setnow) uses the date of the database and (Save Current Date/Time) uses the date/time of the server where it is installed?

**Reply by rijkvanzanten** [reply]
The "Save Current Date/Time" option generates the time on the server (eg Node) and inserts it in the DB. The "Set to Now" button in the app just pulls whatever Date.now() is in the browser, and uses that as the value.

One thing we could try is replacing the server time generation here:

https://github.com/directus/directus/blob/eddc839606d77677818a14996b7e7ee06ae8cc50/api/src/services/payload.ts#L105-L112

to rely on either `knex.fn.now()` (so it uses "NOW()" in the DB instead) or have it use a formatted string that explicitly strips out the timezone (eg `yyyy-MM-dd HH:MM:ss`) to make sure the `tedious` connection package doesn't get confused

**Reply by hugopvl** [reply]
Sounds like a great solution...

**Reply by rijkvanzanten** [reply]
The problem with that second solution is that not every database driver uses the same date format 🙂 🔫 

(feel my pain yet? hehe)

**Reply by hugopvl** [reply]
Yes I feel your pain, it's deep :)

But still, the best solution was to have an option in the app so we could select the timezone!
Or is it out of the question?
Because this way it is more difficult to get around, since we have the TZ of the DB, the TZ Directus Server... and also the TZ of the user :)

**Reply by rijkvanzanten** [reply]
> But still, the best solution was to have an option in the app so we could select the timezone!
> Or is it out of the question?
> Because this way it is more difficult to get around, since we have the TZ of the DB, the TZ Directus Server... and also the TZ of the user :)

Honestly this! If we were to add a third timezone into the mix, before fixing these little quirks between the first two first, thing will get extra difficult

**Reply by Pachat** [reply]
Just to mention that, at least with SQLite,
    `SELECT datetime()` netiher
    `SELECT datetime(strftime('%s'), 'unixepoch'); ` and
    the Directus Datetime widget  do not apply the daylight saving time

but
    `SELECT datetime(strftime('%s'), 'unixepoch', 'localtime'); ` does apply the daylight saving time and gives the correct time

On the JavaScript side
    `Date()` applies the daylight saving time and gives the correct time

--
Another point
> Monday is the first day of the week according to the international standard ISO 8601, 
but in the US, Canada, and Japan, it's counted as the second day of the week

it looks like the Directus Datetime widget  is set to Sunday for everyone

**Reply by rijkvanzanten** [reply]
A tweak to properly fix these timezone inconsistencies was just merged in https://github.com/directus/directus/pull/10956 🙂
